### PR TITLE
[WIP] Update HMC adaptation interface, remove spl.info, and more tests

### PR DESCRIPTION
**Pull request by xukai92**
TODOs
- [x] Log statistics during adaptation
  - E.g. step size, accept rate, etc
- [x] Run MH as a baseline for sanity check on ESS
- [x] Adaptation interface
- [x] Write scripts to test each of below
  - HMC + DA
  - HMC + PD
  - HMC + DA + PD
  - NUTS + DA + PD
- Test cases:
  - [x] gdemo
  - ~~m12.6~~
  - [x] m12.2
  - ~~LDA~~

### CODE REVIEW

File: src/samplers/hmc.jl [hunk 1]
```diff
@@ -267,7 +270,9 @@ function step(model, spl::Sampler{<:Hamiltonian}, vi::VarInfo, is_first::Val{fal
     end
 
     if spl.alg isa AdaptiveHamiltonian
-        adapt!(spl.info[:wum], α, vi[spl], adapt_M=false, adapt_ϵ=true)
+        vi2 = deepcopy(vi)
```
Comment (xukai92) [inline src/samplers/hmc.jl]: @yebai The main change is here.

**Comment by xukai92** [comment]
@yebai I'm currently testing this using the model below to test the adaptation. FYI, as we discussed before, the step size became very small (order of 1e-5) once the pre-condition matrix starts to be used in the first window.

```julia
using StatisticalRethinking
using Turing

Turing.setadbackend(:reverse_diff)

d = CSV.read(joinpath(dirname(Base.pathof(StatisticalRethinking)), "..", "data",
    "reedfrogs.csv"), delim=';')
size(d) # Should be 48x5

# Set number of tanks
d[:tank] = 1:size(d,1)

@model m12_2(density, tank, surv) = begin
    # Separate priors on α and σ for each tank
    σ ~ Truncated(Cauchy(0, 1), 0, Inf)
    α ~ Normal(0, 1) #Check stancode this might need to be in the for loop below?

    # Number of unique tanks in the data set
    N_tank = length(tank)

    # Set an TArray for the priors/param
    # a_tank = TArray{Any}(undef, N_tank)
    a_tank = Vector{Real}(undef, N_tank)

    # For each tank [1,..,48] set a prior
    for i = 1:N_tank
        a_tank[i] ~ Normal(α, σ)
    end

    for i = 1:N_tank
        logitp = a_tank[tank[i]]
        surv[i] ~ BinomialLogit(density[i], logitp)
    end
end

posterior = sample(m12_2(d[:density], d[:tank], d[:surv]),
                   Turing.NUTS(4000, 500, 0.95),
                   # Turing.HMCDA(4000, 500, 0.95, 0.3)
                  )
describe(posterior)
```

Besides the step-size, another metric I was looking at is the ESS, which is super small (<100) for most of the variables.

**Comment by yebai** [comment]
@xukai92  I fixed some bugs in the momentum sampler. See the code for details. The `m12_2` works well on my local machine now and produces much better ESS. 


```julia
julia> posterior = sample(m12_2(d[:density], d[:tank], d[:surv]),
                          Turing.NUTS(4000, 500, 0.95),
                          # Turing.HMCDA(4000, 500, 0.95, 0.3)
                         )
[ Info: [Turing] looking for good initial eps...
[NUTS{Turing.FluxTrackerAD,Union{}}] found initial ϵ: 0.1
[ Info:  Adapted ϵ = 0.1359504498488957, std = [0.0178133, 0.0178358, 0.188375, 0.110931, 0.051245, 0.121746, 0.0784662, 0.0592021, 0.0944426, 0.127033, 0.0493565, 0.13549, 0.0340305, 0.0745223, 0.0445647, 0.0676575, 0.152317, 0.182675, 0.205154, 0.066567, 0.0357704, 0.114956, 0.033635, 0.0343301, 0.0307416, 0.0325884, 0.0218524, 0.012361, 0.0427885, 0.0190135, 0.0225702, 0.0143705, 0.0186903, 0.0104403, 0.0965284, 0.0385374, 0.0313034, 0.0296088, 0.0153678, 0.210476, 0.0494165, 0.0266711, 0.0292146, 0.0101339, 0.00794162, 0.00951529, 0.0106865, 0.00876222, 0.0376052, 0.0167227]; 500 iterations is used for adaption.
[NUTS] Sampling...100% Time: 0:05:57
[NUTS] Finished with
  Mean alpha         = 0.9688205433521577;
  Running time        = 356.5020896809998;
  #lf / sample        = 0.0005;
  #evals / sample     = 73.54075;
  pre-cond. metric    = [0.0178133, 0.0178358, 0.18837....


julia> describe(posterior)
Iterations = 1:4000
Thinning interval = 1
Chains = 1
Samples per chain = 4000

Empirical Posterior Estimates:
                Mean           SD         Naive SE       MCSE         ESS    
         α    1.169201998  0.050174783 0.00079333298 0.0026625587  355.117953
a_tank[33]    2.791068028  0.172694599 0.00273054136 0.0115638423  223.024817
 a_tank[2]    2.440550622  0.211548960 0.00334488275 0.0165119221  164.144573
a_tank[44]   -0.294770345  0.049875325 0.00078859813 0.0014100555 1251.120277
a_tank[41]   -1.616970703  0.089028596 0.00140766570 0.0037296095  569.812965
a_tank[40]    2.157485261  0.092159753 0.00145717363 0.0055486929  275.867933
a_tank[42]   -0.515417817  0.066885741 0.00105755643 0.0014340261 2175.469514
a_tank[11]    0.926972707  0.109216405 0.00172686298 0.0022205224 2419.164594
a_tank[16]    1.812176198  0.207088917 0.00327436329 0.0097857221  447.845238
a_tank[46]   -0.515145458  0.076731062 0.00121322462 0.0011681490 4000.000000
 a_tank[8]    1.814482790  0.179746698 0.00284204483 0.0098495774  333.032519
a_tank[14]    0.238785246  0.130159757 0.00205800646 0.0044540900  853.955965
         σ    1.246601073  0.078598935 0.00124275828 0.0098855761   63.216341
a_tank[47]    1.915402384  0.093933510 0.00148521920 0.0024131698 1515.183841
a_tank[43]   -0.401474816  0.083113079 0.00131413316 0.0006679581 4000.000000
a_tank[24]    1.579917860  0.094669134 0.00149685044 0.0041848579  511.747076
a_tank[34]    2.437256173  0.116599652 0.00184360238 0.0049195221  561.757246
 a_tank[9]   -0.074912991  0.095801893 0.00151476093 0.0026630504 1294.164320
a_tank[28]   -0.399940492  0.115927322 0.00183297190 0.0022362573 2687.373762
a_tank[31]   -0.556711350  0.063110270 0.00099786099 0.0026034027  587.648181
a_tank[29]    0.181933660  0.072268375 0.00114266334 0.0019324941 1398.492822
a_tank[18]    2.139981467  0.146485800 0.00231614387 0.0092294565  251.906165
 a_tank[1]    1.806432336  0.228248835 0.00360893096 0.0097725515  545.508099
a_tank[38]    3.274080572  0.337149649 0.00533080401 0.0300958730  125.496476
        lp -153.459963218  6.762417305 0.10692320587 0.8360738534   65.420619
a_tank[37]    1.917151354  0.071389090 0.00112876063 0.0020628797 1197.611363
a_tank[30]    1.357187750  0.075334566 0.00119114408 0.0015833823 2263.690230
    lf_eps    0.126918124  0.034771383 0.00054978383 0.0042624621   66.546139
a_tank[35]    2.440524668  0.131122219 0.00207322432 0.0071111899  339.991474
a_tank[20]    3.002449787  0.236246288 0.00373538180 0.0146859243  258.778040
a_tank[15]    1.805615524  0.199484992 0.00315413467 0.0076413208  681.527818
   epsilon    0.126918124  0.034771383 0.00054978383 0.0042624621   66.546139
a_tank[26]    0.184759576  0.068118443 0.00107704715 0.0009244528 4000.000000
a_tank[45]    0.573788829  0.064236795 0.00101567291 0.0009784361 4000.000000
a_tank[21]    2.146172803  0.115106758 0.00181999765 0.0054473561  446.508891
  eval_num   73.540750000 38.365900550 0.60661815111 4.3702545016   77.068576
a_tank[48]    0.025890056  0.068374197 0.00108109099 0.0031354030  475.551708
a_tank[10]    1.827566094  0.211661141 0.00334665648 0.0160431473  174.061661
 a_tank[3]    0.928380130  0.112241436 0.00177469293 0.0032264046 1210.232405
 a_tank[7]    2.428742522  0.211991674 0.00335188267 0.0156443007  183.622246
    lf_num    0.000500000  0.031622777 0.00050000000 0.0005000000 4000.000000
 a_tank[6]    1.815136370  0.160873309 0.00254363035 0.0093616979  295.296799
a_tank[25]   -0.880870192  0.104817856 0.00165731583 0.0032750251 1024.333303
a_tank[32]   -0.254668649  0.102768023 0.00162490512 0.0015839325 4000.000000
 a_tank[5]    1.814822277  0.161158829 0.00254814483 0.0100584762  256.710614
a_tank[13]    0.928501968  0.102365982 0.00161854830 0.0024996998 1677.009767
   elapsed    0.089125522  0.073228688 0.00115784722 0.0054541410  180.264322
a_tank[19]    1.838526051  0.113785388 0.00179910495 0.0043036090  699.048602
a_tank[12]    0.576835853  0.111622956 0.00176491390 0.0045880526  591.902632
 a_tank[4]    2.432091015  0.220041390 0.00347915986 0.0165270942  177.262044
a_tank[17]    2.520300546  0.202604469 0.00320345793 0.0093532139  469.219774
a_tank[27]   -1.252784872  0.107965330 0.00170708175 0.0047200410  523.211270
a_tank[22]    2.142229515  0.111551500 0.00176378409 0.0089861654  154.099778
a_tank[23]    2.144140377  0.109617333 0.00173320221 0.0063939121  293.917296
a_tank[36]    1.924087203  0.111087965 0.00175645495 0.0058388853  361.971303
a_tank[39]    2.442887608  0.143394433 0.00226726506 0.0091175876  247.345888

```

**Comment by yebai** [comment]
Updated numbers produced using 29cd4c8

```julia

[ Info: [Turing] looking for good initial eps...
[NUTS{Turing.Core.FluxTrackerAD,Union{}}] found initial ϵ: 0.1
[ Info:  Adapted ϵ = 0.09337172168698085, std = [0.0166337, 0.0213772, 0.0848869, 0.12702, 0.102382, 0.461173, 0.0828356, 0.210166, 0.162271, 0.100355, 0.0598037, 0.189802, 0.14314, 0.0609404, 0.0632079, 0.0560756, 0.153805, 0.202865, 0.0902155, 0.117504, 0.0351577, 0.185393, 0.0642132, 0.0530938, 0.0343522, 0.0362635, 0.0196606, 0.0197741, 0.0387709, 0.0192463, 0.0125843, 0.0271536, 0.0126304, 0.0141961, 0.127015, 0.0613429, 0.0668559, 0.023086, 0.0344057, 0.101323, 0.0331063, 0.0398331, 0.0200233, 0.0084846, 0.0159985, 0.0097223, 0.0234096, 0.0101566, 0.0543068, 0.0145356]; 500 iterations is used for adaption.
[NUTS] Sampling...100% Time: 0:08:34
[NUTS] Finished with
  Mean alpha         = 0.9478608859909594;
  Running time        = 513.188298636999;
  #lf / sample        = 0.0005;
  #evals / sample     = 101.269;
  pre-cond. metric    = [0.0166337, 0.0213772, 0.08488....

julia> describe(posterior)
Iterations = 1:4000
Thinning interval = 1
Chains = 1
Samples per chain = 4000

Empirical Posterior Estimates:
                Mean           SD         Naive SE        MCSE         ESS    
         α    1.169280740  0.045350297 0.00071705116 0.00251274924  325.733165
a_tank[33]    2.786162558  0.170514211 0.00269606639 0.00801311495  452.812505
 a_tank[2]    2.430304649  0.281284153 0.00444749296 0.01787880021  247.521990
a_tank[44]   -0.292619425  0.065812512 0.00104058718 0.00085047095 4000.000000
a_tank[41]   -1.613134375  0.082613998 0.00130624200 0.00241581132 1169.448911
a_tank[40]    2.158473227  0.092038533 0.00145525698 0.00486760316  357.527163
a_tank[42]   -0.514625899  0.063406216 0.00100254030 0.00052846471 4000.000000
a_tank[11]    0.921824461  0.162243174 0.00256528982 0.00741044753  479.340356
a_tank[16]    1.807393322  0.235618071 0.00372544881 0.00696930663 1142.978457
a_tank[46]   -0.514419171  0.076794633 0.00121422976 0.00049234203 4000.000000
 a_tank[8]    1.814814526  0.169665903 0.00268265347 0.00856247803  392.636079
a_tank[14]    0.247201440  0.112170922 0.00177357801 0.00421818595  707.146083
         σ    1.247065050  0.068229831 0.00107880835 0.00913256280   55.816585
a_tank[47]    1.918853447  0.100366714 0.00158693708 0.00290813920 1191.102197
a_tank[43]   -0.400580780  0.074476263 0.00117757312 0.00203470056 1339.783880
a_tank[24]    1.586094672  0.088514859 0.00139954281 0.00380714512  540.547049
a_tank[34]    2.444549934  0.108498211 0.00171550735 0.00832493190  169.857128
 a_tank[9]   -0.072817629  0.112679784 0.00178162382 0.00326784706 1188.963625
a_tank[28]   -0.404929306  0.082175045 0.00129930155 0.00277181295  878.925494
a_tank[31]   -0.555112028  0.079640621 0.00125922879 0.00169546290 2206.442903
a_tank[29]    0.184233131  0.062754649 0.00099223812 0.00131847319 2265.422453
a_tank[18]    2.149969553  0.135771477 0.00214673554 0.00712329912  363.291067
 a_tank[1]    1.809505734  0.143882816 0.00227498707 0.00944012403  232.307085
a_tank[38]    3.231800199  0.202456098 0.00320111198 0.01558832349  168.679733
        lp -153.500944720  6.590704713 0.10420819139 0.77623736375   72.089933
a_tank[37]    1.916741085  0.101602485 0.00160647635 0.00308673756 1083.450939
a_tank[30]    1.356192247  0.072043774 0.00113911209 0.00096548497 4000.000000
    lf_eps    0.086146408  0.028573389 0.00045178495 0.00329739640   75.089842
a_tank[35]    2.436980653  0.123229430 0.00194842837 0.00654520823  354.472008
a_tank[20]    3.019337193  0.228852311 0.00361847276 0.01415361362  261.442348
a_tank[15]    1.816680411  0.178491597 0.00282219994 0.00698982351  652.083375
   epsilon    0.086146408  0.028573389 0.00045178495 0.00329739640   75.089842
a_tank[26]    0.182692857  0.067775390 0.00107162301 0.00122337900 3069.177515
a_tank[45]    0.571693300  0.083454393 0.00131952982 0.00231892729 1295.162158
a_tank[21]    2.142004118  0.108385889 0.00171373138 0.00409394087  700.910168
  eval_num  101.269000000 30.394343109 0.48057676104 4.17473419697   53.006346
a_tank[48]    0.027221487  0.062474208 0.00098780397 0.00160825383 1509.010744
a_tank[10]    1.798121834  0.223204078 0.00352916635 0.01088459171  420.513523
 a_tank[3]    0.933724876  0.144733321 0.00228843474 0.00835004325  300.441282
 a_tank[7]    2.428258346  0.204971828 0.00324088916 0.00807774729  643.884281
    lf_num    0.000500000  0.031622777 0.00050000000 0.00050000000 4000.000000
 a_tank[6]    1.815309605  0.223521137 0.00353417948 0.01033435368  467.811133
a_tank[25]   -0.879727184  0.094501956 0.00149420711 0.00143064627 4000.000000
a_tank[32]   -0.252255567  0.086950086 0.00137480157 0.00257518172 1140.050971
 a_tank[5]    1.813432533  0.149109754 0.00235763222 0.00771835514  373.218011
a_tank[13]    0.931725354  0.111324484 0.00176019465 0.00553587705  404.397545
   elapsed    0.128297075  0.068768749 0.00108732939 0.00557990732  151.889683
a_tank[19]    1.841636652  0.123363729 0.00195055182 0.00382407785 1040.690353
a_tank[12]    0.567265284  0.101549806 0.00160564342 0.00214312781 2245.236300
 a_tank[4]    2.443585242  0.419414526 0.00663152593 0.00925108930 2055.422889
a_tank[17]    2.527348400  0.156595599 0.00247599383 0.01205409137  168.768017
a_tank[27]   -1.252521861  0.105848836 0.00167361705 0.00324303507 1065.292536
a_tank[22]    2.138817535  0.115114960 0.00182012733 0.00216839706 2818.291527
a_tank[23]    2.141540237  0.086613346 0.00136947725 0.00279862097  957.814635
a_tank[36]    1.919089511  0.083828634 0.00132544709 0.00345601155  588.348178
a_tank[39]    2.435379041  0.094117587 0.00148812972 0.00603547183  243.175106
```

**Comment by xukai92** [comment]
@fisherc2 Thanks for the attention. I will include m12.6 as our testing example as well. You're right here the running time might be related to this bug because if the step-size is tuned badly NUTS could run a lot of steps when building the tree. Thanks!

**Comment by xukai92** [comment]
TODOs
- [ ] Run MH as a baseline for sanity check on ESS

**Comment by xukai92** [comment]
I reverted https://github.com/TuringLang/Turing.jl/commit/16d6bba492fb4c5ee347874f2360cb4e681b2bee temporarily to let tests pass. I will update the interfaces first so that it can be easily experimented. Then we can check that later.

**Comment by xukai92** [comment]
@yebai When designing the interface for setting configuration for adaptation, I realised that we need to tie adaptation algorithm with sampling algorithm, otherwise the Gibbs interface is hard to work with.

**Comment by trappmartin** [comment]
I started playing around with the NUTS sampler trying to improve the performance. So far I didn’t had the time to make systemic tests on my changes. I’ll evaluate and possibly integrate the changes after this PR is merged.

Do you also plan to look at ignoring the samples from the adaption phase for the returned chain object?

**Comment by yebai** [comment]
> I reverted [16d6bba](https://github.com/TuringLang/Turing.jl/commit/16d6bba492fb4c5ee347874f2360cb4e681b2bee) temporarily to let tests pass. I will update the interfaces first so that it can be easily experimented. Then we can check that later.

@xukai92 
Sorry for the confusion. This PR contains some temporary code to adapt the pre-conditioning matrix in the original space, instead of the transformed space  (i.e. Euclidean). This is no longer necessary after 16d6bba492fb4c5ee347874f2360cb4e681b2bee. Pls take a look at: https://github.com/TuringLang/Turing.jl/commit/29cd4c85bc1c0bb022ee44838a793947ba9bd6e1

**Comment by xukai92** [comment]
Yes this is the current setting I'm experimenting, i.e. the original v.s. Euclidean space codes are removed (as in https://github.com/TuringLang/Turing.jl/commit/29cd4c85bc1c0bb022ee44838a793947ba9bd6e1). But with the modification in https://github.com/TuringLang/Turing.jl/commit/16d6bba492fb4c5ee347874f2360cb4e681b2bee, the inference results doesn't seem to be correct, e.g. the inference check of `gdemo` in `test/nuts.jl/nuts.jl` doesn't pass with diagonal pre-conditioner.

**Comment by yebai** [comment]
> Yes this is the current setting I'm experimenting, i.e. the original v.s. Euclidean space codes are removed (as in [29cd4c8](https://github.com/TuringLang/Turing.jl/commit/29cd4c85bc1c0bb022ee44838a793947ba9bd6e1)). But with the modification in [16d6bba](https://github.com/TuringLang/Turing.jl/commit/16d6bba492fb4c5ee347874f2360cb4e681b2bee), the inference results doesn't seem to be correct, e.g. the inference check of `gdemo` in `test/nuts.jl/nuts.jl` doesn't pass with diagonal pre-conditioner.

Hi Kai, I went through the code again and your version of momentum is correct. I was confused by the covariance of theta, and the covariance of momentum variables. However, it's strange the incorrect momentum gives better ESS. Any idea?

**Comment by xukai92** [comment]
Correct me if I'm wrong: I think what happened was the incorrect momentum somehow resulted in a large step size after adaptation phase. Running NUTS with larger step size which would just lead to larger ESS regardless of whether or not the momentum sampler is correct.

Also the table seems to indicate that there is something wrong within NUTS alone, because with diagonal PC, other samplers do have a gain in ESS.

I looked into the NUTS code again and found these lines suspicious. I think we should also correct momentum variable by PC matrix here: https://github.com/TuringLang/Turing.jl/blob/kx/fix-pd-bug/src/inference/nuts.jl#L117 and https://github.com/TuringLang/Turing.jl/blob/kx/fix-pd-bug/src/inference/nuts.jl#L174

Also I haven't seen ESS gain by doing so neither, I believe this should be fixed.

**Comment by yebai** [comment]
Updated ESS using 300dda3


```julia

julia> posterior = sample(m12_2(d[:density], d[:tank], d[:surv]),
                          Turing.NUTS(4000, 500, 0.95),
                          # Turing.HMCDA(4000, 500, 0.95, 0.3)
                         )
[ Info: [Turing] looking for good initial eps...
[ Info: [Turing] eps = 0.1, accept ratio a = 1.1184742546043598
[ Info: [Turing] eps = 0.11000000000000001, accept ratio a = 1.1942888456178937
[ Info: [Turing] eps = 0.12100000000000002, accept ratio a = 1.3234513576835787
[ Info: [Turing] eps = 0.13310000000000002, accept ratio a = 1.555321884953187
[ Info: [Turing] eps = 0.14641000000000004, accept ratio a = 2.008758175882085
[ Info: [Turing] eps = 0.16105100000000006, accept ratio a = 3.029480023216812
[ Info: [Turing] eps = 0.17715610000000007, accept ratio a = 5.942575359627319
[ Info: [Turing] eps = 0.1948717100000001, accept ratio a = 18.5796781383723
[ Info: [Turing] eps = 0.2143588810000001, accept ratio a = 139.84101172263613
[ Info: [Turing] eps = 0.23579476910000013, accept ratio a = 6286.38203235946
[ Info: [Turing] eps = 0.25937424601000014, accept ratio a = 1.4602934555244546e7
[ Info: [Turing] eps = 0.28531167061100016, accept ratio a = 2.927396270806854e14
[ Info: [Turing] eps = 0.3138428376721002, accept ratio a = 3.2374057724820245e28
[NUTS] Sampling...100% Time: 0:06:56
[NUTS] Finished with
  Running time        = 415.0562515230004;
  #lf / sample        = 0.0;
  #evals / sample     = 88.89675;
  pre-cond. metric    = [1.0, 1.0, 1.0, 1.0, 1.0, 1.0,....
Object of type Chains, with data of type 4000×56×1 Array{Union{Missing, Float64},3}

Log evidence      = 0.0
Iterations        = 1:4000
Thinning interval = 1
Chains            = Chain1
Samples per chain = 4000
internals         = elapsed, epsilon, eval_num, lf_eps, lf_num, lp
parameters        = a_tank[25], a_tank[41], a_tank[18], a_tank[1], a_tank[27], a_tank[7], a_tank[15], a_tank[12], a_tank[6], a_tank[31], a_tank[26], a_tank[11], a_tank[8], a_tank[43], a_tank[48], a_tank[10], a_tank[20], a_tank[45], a_tank[47], a_tank[21], a_tank[37], a_tank[39], a_tank[13], a_tank[9], a_tank[29], a_tank[38], a_tank[40], a_tank[34], a_tank[19], a_tank[2], a_tank[14], σ, a_tank[36], a_tank[22], a_tank[5], a_tank[4], α, a_tank[42], a_tank[23], a_tank[35], a_tank[46], a_tank[32], a_tank[30], a_tank[17], a_tank[24], a_tank[33], a_tank[44], a_tank[28], a_tank[3], a_tank[16]

parameters
             Mean    SD   Naive SE  MCSE     ESS   
 a_tank[1]  2.0874 0.8569   0.0135 0.0184 2179.6495
 a_tank[2]  3.0820 1.1454   0.0181 0.0273 1759.0694
 a_tank[3]  0.9737 0.6478   0.0102 0.0098 4000.0000
 a_tank[4]  3.0146 1.1085   0.0175 0.0254 1902.7665
 a_tank[5]  2.1342 0.8555   0.0135 0.0137 3910.0172
 a_tank[6]  2.1370 0.8716   0.0138 0.0218 1595.1616
 a_tank[7]  3.0481 1.1157   0.0176 0.0320 1214.4261
 a_tank[8]  2.1155 0.8624   0.0136 0.0170 2571.8287
 a_tank[9] -0.1722 0.6021   0.0095 0.0151 1583.3423
a_tank[10]  2.1045 0.8533   0.0135 0.0180 2243.0769
a_tank[11]  0.9942 0.6701   0.0106 0.0103 4000.0000
a_tank[12]  0.5854 0.6288   0.0099 0.0092 4000.0000
a_tank[13]  0.9743 0.6388   0.0101 0.0102 3909.9762
a_tank[14]  0.1790 0.6267   0.0099 0.0082 4000.0000
a_tank[15]  2.1105 0.8666   0.0137 0.0146 3539.0543
a_tank[16]  2.1146 0.8622   0.0136 0.0190 2056.2411
a_tank[17]  2.9000 0.7873   0.0124 0.0135 3387.2094
a_tank[18]  2.3901 0.6683   0.0106 0.0115 3374.4636
a_tank[19]  2.0135 0.5798   0.0092 0.0102 3200.4977
a_tank[20]  3.6400 1.0313   0.0163 0.0276 1394.2323
a_tank[21]  2.3826 0.6745   0.0107 0.0125 2907.9989
a_tank[22]  2.3759 0.6647   0.0105 0.0140 2254.5778
a_tank[23]  2.3883 0.6628   0.0105 0.0111 3559.9914
a_tank[24]  1.7034 0.5300   0.0084 0.0069 4000.0000
a_tank[25] -1.0049 0.4489   0.0071 0.0101 1975.0659
a_tank[26]  0.1652 0.3914   0.0062 0.0063 3870.3732
a_tank[27] -1.4201 0.5107   0.0081 0.0146 1219.1459
a_tank[28] -0.4797 0.3978   0.0063 0.0076 2758.4433
a_tank[29]  0.1706 0.3956   0.0063 0.0042 4000.0000
a_tank[30]  1.4346 0.5148   0.0081 0.0122 1782.7890
a_tank[31] -0.6248 0.4405   0.0070 0.0099 1970.4885
a_tank[32] -0.3114 0.3986   0.0063 0.0082 2388.5491
a_tank[33]  3.1679 0.7608   0.0120 0.0134 3209.3237
a_tank[34]  2.6963 0.6204   0.0098 0.0106 3438.3232
a_tank[35]  2.7023 0.6481   0.0102 0.0127 2589.1435
a_tank[36]  2.0551 0.5070   0.0080 0.0096 2815.1312
a_tank[37]  2.0538 0.5110   0.0081 0.0060 4000.0000
a_tank[38]  3.8388 0.9664   0.0153 0.0223 1873.4088
a_tank[39]  2.7135 0.6457   0.0102 0.0151 1834.0260
a_tank[40]  2.3467 0.5586   0.0088 0.0099 3176.7577
a_tank[41] -1.8039 0.4903   0.0078 0.0101 2354.3516
a_tank[42] -0.5749 0.3446   0.0054 0.0052 4000.0000
a_tank[43] -0.4566 0.3473   0.0055 0.0069 2539.3825
a_tank[44] -0.3335 0.3584   0.0057 0.0083 1869.6294
a_tank[45]  0.5716 0.3429   0.0054 0.0057 3681.3659
a_tank[46] -0.5607 0.3534   0.0056 0.0054 4000.0000
a_tank[47]  2.0666 0.5149   0.0081 0.0078 4000.0000
a_tank[48]  0.0037 0.3413   0.0054 0.0063 2943.2044
         α  1.2902 0.2421   0.0038 0.0038 3989.0677
         σ  1.6135 0.2246   0.0036 0.0067 1125.1501
```

**Comment by xukai92** [comment]
So does it mean it's the initial step size was too bad?

**Comment by yebai** [comment]
yes; there is another subtle bug fixed by https://github.com/TuringLang/Turing.jl/pull/709/commits/cdf7e34911a0bae680ce4b69836d873ee58c43c0, which is the possible cause of turning stepsize to extreme values occationally.

**Comment by xukai92** [comment]
Got it. That's great! Thanks for catching that!

**Comment by yebai** [comment]
Updated ESS using 8486578

```julia
 julia> describe(c[1])
Log evidence      = 0.0
Iterations        = 1:4000
Thinning interval = 1
Chains            = 1
Samples per chain = 4000
parameters        = a_tank[25], a_tank[41], a_tank[18], a_tank[1], a_tank[27], a_tank[7], a_tank[15], a_tank[12], a_tank[6], a_tank[31], a_tank[26], a_tank[11], a_tank[8], a_tank[43], a_tank[48], a_tank[10], a_tank[20], a_tank[45], a_tank[47], a_tank[21], a_tank[37], a_tank[39], a_tank[13], a_tank[9], a_tank[29], a_tank[38], a_tank[40], a_tank[34], a_tank[19], a_tank[2], a_tank[14], σ, a_tank[36], a_tank[22], a_tank[5], a_tank[4], α, a_tank[42], a_tank[23], a_tank[35], a_tank[46], a_tank[32], a_tank[30], a_tank[17], a_tank[24], a_tank[33], a_tank[44], a_tank[28], a_tank[3], a_tank[16]

Empirical Posterior Estimates
───────────────────────────────────────────────────
parameters
             Mean    SD   Naive SE  MCSE     ESS   
 a_tank[1]  2.1003 0.8569   0.0135 0.0134 4000.0000
 a_tank[2]  3.0777 1.1224   0.0177 0.0206 2966.8863
 a_tank[3]  1.0122 0.7028   0.0111 0.0081 4000.0000
 a_tank[4]  3.0292 1.1208   0.0177 0.0256 1910.1714
 a_tank[5]  2.1234 0.8508   0.0135 0.0128 4000.0000
 a_tank[6]  2.1137 0.8874   0.0140 0.0141 3933.2352
 a_tank[7]  3.0676 1.1364   0.0180 0.0276 1699.9622
 a_tank[8]  2.1202 0.8758   0.0138 0.0149 3471.2253
 a_tank[9] -0.1771 0.6253   0.0099 0.0067 4000.0000
a_tank[10]  2.1044 0.8639   0.0137 0.0126 4000.0000
a_tank[11]  0.9947 0.6780   0.0107 0.0098 4000.0000
a_tank[12]  0.5707 0.6176   0.0098 0.0049 4000.0000
a_tank[13]  0.9857 0.6648   0.0105 0.0091 4000.0000
a_tank[14]  0.1955 0.6165   0.0097 0.0074 4000.0000
a_tank[15]  2.1081 0.8577   0.0136 0.0150 3286.1862
a_tank[16]  2.1171 0.8639   0.0137 0.0130 4000.0000
a_tank[17]  2.8987 0.8101   0.0128 0.0132 3751.5649
a_tank[18]  2.3839 0.6712   0.0106 0.0091 4000.0000
a_tank[19]  2.0169 0.5924   0.0094 0.0075 4000.0000
a_tank[20]  3.6290 0.9808   0.0155 0.0200 2410.0861
a_tank[21]  2.3844 0.6692   0.0106 0.0094 4000.0000
a_tank[22]  2.3876 0.6573   0.0104 0.0095 4000.0000
a_tank[23]  2.3754 0.6857   0.0108 0.0110 3902.7362
a_tank[24]  1.7012 0.5532   0.0087 0.0069 4000.0000
a_tank[25] -0.9958 0.4517   0.0071 0.0062 4000.0000
a_tank[26]  0.1580 0.3855   0.0061 0.0058 4000.0000
a_tank[27] -1.4286 0.4989   0.0079 0.0070 4000.0000
a_tank[28] -0.4718 0.4123   0.0065 0.0065 4000.0000
a_tank[29]  0.1631 0.4030   0.0064 0.0055 4000.0000
a_tank[30]  1.4453 0.4935   0.0078 0.0067 4000.0000
a_tank[31] -0.6408 0.4352   0.0069 0.0057 4000.0000
a_tank[32] -0.3140 0.3977   0.0063 0.0045 4000.0000
a_tank[33]  3.1940 0.7839   0.0124 0.0139 3191.3482
a_tank[34]  2.7118 0.6569   0.0104 0.0104 4000.0000
a_tank[35]  2.7100 0.6370   0.0101 0.0104 3768.0450
a_tank[36]  2.0645 0.5210   0.0082 0.0066 4000.0000
a_tank[37]  2.0538 0.5074   0.0080 0.0074 4000.0000
a_tank[38]  3.8684 0.9732   0.0154 0.0200 2365.1137
a_tank[39]  2.6992 0.6491   0.0103 0.0092 4000.0000
a_tank[40]  2.3446 0.5462   0.0086 0.0079 4000.0000
a_tank[41] -1.8085 0.4845   0.0077 0.0061 4000.0000
a_tank[42] -0.5761 0.3455   0.0055 0.0048 4000.0000
a_tank[43] -0.4552 0.3472   0.0055 0.0047 4000.0000
a_tank[44] -0.3321 0.3497   0.0055 0.0051 4000.0000
a_tank[45]  0.5760 0.3511   0.0056 0.0052 4000.0000
a_tank[46] -0.5664 0.3604   0.0057 0.0050 4000.0000
a_tank[47]  2.0663 0.5150   0.0081 0.0077 4000.0000
a_tank[48] -0.0036 0.3386   0.0054 0.0051 4000.0000
         α  1.2925 0.2536   0.0040 0.0042 3676.5900
         σ  1.6180 0.2251   0.0036 0.0049 2151.3881
```

**Comment by mohamed82008** [comment]
How does this PR relate to AdvancedHMC?

**Comment by xukai92** [comment]
@mohamed82008 This PR was originally created to fix some adaptation numerical bugs and update Turing's interface to pass different adaptation options. However, as 
1. @yebai fixed the main bugs in another PR
2. some interface issues are resolved by AdvancedHMC
the new codes introduced by this PR is very limited - mainly it's just remove all in-sampler statistics from spl.info and instead returning them by each `step` function. To achieve this I introduced a new type for each sampler. E.g. `HMCState` for HMC that contains stats collected during each HMC iteration, to replace the original returned accepted probability (https://github.com/TuringLang/Turing.jl/pull/641/files#diff-b2000c195c9a7dd900b574dec178b527R224). 

In fact, I'm thinking closing this PR first and do a new one later, when AdvancedHMC is stable to be integrated. Otherwise it's a bit tricky to maintain two copies of codes for the same functionality. How do you think?

**Comment by mohamed82008** [comment]
Yes, currently I prefer minimal code addition to Turing. Code removal is fine. So if bug fixes already went into Turing, I recommend waiting for AdvancedHMC.jl. I finally have a working state of the type stability PR (at least for Hamiltonian), really not looking forward to a 3-hour rebase!

**Comment by xukai92** [comment]
Nice! Neither do I was enjoying resolving all the conflicts yesterday :( Let's close it then

**Comment by mohamed82008** [comment]
Thanks!

**Pull request closed without merging**